File: lib/actions/srs.ts
"use server";

import { db } from "@/lib/db";
import { srsCard } from "@/lib/db/schema";
import { and, eq, lte, gt, count, sql, asc, isNotNull, inArray } from "drizzle-orm";
import { requireSession } from "@/lib/auth-server";
import { calculateNextReview, type Quality, type CardStatus } from "@/lib/srs";

export async function addWordToSrs(
  word: string,
  language: string,
  translation: string
) {
  const session = await requireSession();

  await db
    .insert(srsCard)
    .values({
      word: word.toLowerCase(),
      language,
      userId: session.user.id,
      translation,
      status: "new",
      nextReviewAt: null,
    })
    .onConflictDoNothing();
}

/**
 * Add a word to SRS if new, or mark it as failed to remember if it already exists.
 * New words go to "learning" with nextReviewAt = now (tooltip path — user is actively studying).
 * Returns "added" if the word was newly added, "failed" if it was reset.
 */
export async function addOrFailWord(
  word: string,
  language: string,
  translation: string
): Promise<"added" | "failed"> {
  const session = await requireSession();
  const userId = session.user.id;
  const normalizedWord = word.toLowerCase();

  const [existing] = await db
    .select()
    .from(srsCard)
    .where(
      and(
        eq(srsCard.word, normalizedWord),
        eq(srsCard.language, language),
        eq(srsCard.userId, userId)
      )
    );

  if (!existing) {
    await db.insert(srsCard).values({
      word: normalizedWord,
      language,
      userId,
      translation,
      status: "learning",
      nextReviewAt: new Date(),
    });
    return "added";
  }

  // Already exists — mark as failed to remember (quality = 0)
  const result = calculateNextReview(
    {
      easeFactor: existing.easeFactor,
      interval: existing.interval,
      repetitions: existing.repetitions,
      status: (existing.status as CardStatus) ?? "learning",
    },
    0
  );

  await db
    .update(srsCard)
    .set({
      easeFactor: result.easeFactor,
      interval: result.interval,
      repetitions: result.repetitions,
      status: result.status,
      nextReviewAt: result.nextReviewAt,
      lastReviewedAt: new Date(),
    })
    .where(
      and(
        eq(srsCard.word, normalizedWord),
        eq(srsCard.language, language),
        eq(srsCard.userId, userId)
      )
    );

  return "failed";
}

export async function bulkAddWordsToSrs(
  words: { word: string; translation: string }[],
  language: string
) {
  const session = await requireSession();

  if (words.length === 0) return 0;

  const values = words.map((w) => ({
    word: w.word.toLowerCase(),
    language,
    userId: session.user.id,
    translation: w.translation,
    status: "new" as const,
    nextReviewAt: null,
  }));

  // Insert in batches of 500 to avoid query size limits
  const BATCH_SIZE = 500;
  for (let i = 0; i < values.length; i += BATCH_SIZE) {
    const batch = values.slice(i, i + BATCH_SIZE);
    await db.insert(srsCard).values(batch).onConflictDoNothing();
  }

  return values.length;
}

export async function removeAllWordsFromSrs(language: string) {
  const session = await requireSession();

  await db
    .delete(srsCard)
    .where(
      and(
        eq(srsCard.language, language),
        eq(srsCard.userId, session.user.id)
      )
    );
}

export async function removeWordFromSrs(word: string, language: string) {
  const session = await requireSession();

  await db
    .delete(srsCard)
    .where(
      and(
        eq(srsCard.word, word.toLowerCase()),
        eq(srsCard.language, language),
        eq(srsCard.userId, session.user.id)
      )
    );
}

export async function getDueCards(language?: string, limit = 20) {
  const session = await requireSession();
  const now = new Date();

  const conditions = [
    eq(srsCard.userId, session.user.id),
    inArray(srsCard.status, ["learning", "review"]),
    isNotNull(srsCard.nextReviewAt),
    lte(srsCard.nextReviewAt, now),
  ];

  if (language) {
    conditions.push(eq(srsCard.language, language));
  }

  return db
    .select()
    .from(srsCard)
    .where(and(...conditions))
    .orderBy(srsCard.nextReviewAt)
    .limit(limit);
}

export async function getScheduledCards(language?: string, limit = 20) {
  const session = await requireSession();
  const now = new Date();

  const conditions = [
    eq(srsCard.userId, session.user.id),
    inArray(srsCard.status, ["learning", "review"]),
    isNotNull(srsCard.nextReviewAt),
    gt(srsCard.nextReviewAt, now),
  ];

  if (language) {
    conditions.push(eq(srsCard.language, language));
  }

  return db
    .select()
    .from(srsCard)
    .where(and(...conditions))
    .orderBy(srsCard.nextReviewAt)
    .limit(limit);
}

export async function getAllCards(language?: string) {
  const session = await requireSession();

  const conditions = [eq(srsCard.userId, session.user.id)];

  if (language) {
    conditions.push(eq(srsCard.language, language));
  }

  return db
    .select()
    .from(srsCard)
    .where(and(...conditions))
    .orderBy(srsCard.createdAt);
}

export async function reviewCard(
  word: string,
  language: string,
  quality: Quality
) {
  const session = await requireSession();
  const userId = session.user.id;
  const normalizedWord = word.toLowerCase();

  const [card] = await db
    .select()
    .from(srsCard)
    .where(
      and(
        eq(srsCard.word, normalizedWord),
        eq(srsCard.language, language),
        eq(srsCard.userId, userId)
      )
    );

  if (!card) throw new Error("Card not found");

  const result = calculateNextReview(
    {
      easeFactor: card.easeFactor,
      interval: card.interval,
      repetitions: card.repetitions,
      status: (card.status as CardStatus) ?? "learning",
    },
    quality
  );

  await db
    .update(srsCard)
    .set({
      easeFactor: result.easeFactor,
      interval: result.interval,
      repetitions: result.repetitions,
      status: result.status,
      nextReviewAt: result.nextReviewAt,
      lastReviewedAt: new Date(),
    })
    .where(
      and(
        eq(srsCard.word, normalizedWord),
        eq(srsCard.language, language),
        eq(srsCard.userId, userId)
      )
    );

  return result;
}

export async function getSrsStats(language?: string) {
  const session = await requireSession();
  const now = new Date();

  const conditions = [eq(srsCard.userId, session.user.id)];
  if (language) {
    conditions.push(eq(srsCard.language, language));
  }

  const baseWhere = and(...conditions);

  const [total] = await db
    .select({ count: count() })
    .from(srsCard)
    .where(baseWhere);

  const [due] = await db
    .select({ count: count() })
    .from(srsCard)
    .where(
      and(
        baseWhere,
        inArray(srsCard.status, ["learning", "review"]),
        isNotNull(srsCard.nextReviewAt),
        lte(srsCard.nextReviewAt, now)
      )
    );

  const [newCards] = await db
    .select({ count: count() })
    .from(srsCard)
    .where(and(baseWhere, eq(srsCard.status, "new")));

  const [learning] = await db
    .select({ count: count() })
    .from(srsCard)
    .where(and(baseWhere, eq(srsCard.status, "learning")));

  const [review] = await db
    .select({ count: count() })
    .from(srsCard)
    .where(and(baseWhere, eq(srsCard.status, "review")));

  return {
    total: total.count,
    due: due.count,
    new: newCards.count,
    learning: learning.count,
    review: review.count,
    learned: review.count, // alias for backwards compat
  };
}

export async function getNewCards(language: string, limit = 20) {
  const session = await requireSession();

  return db
    .select()
    .from(srsCard)
    .where(
      and(
        eq(srsCard.userId, session.user.id),
        eq(srsCard.language, language),
        eq(srsCard.status, "new")
      )
    )
    .orderBy(asc(srsCard.createdAt))
    .limit(limit);
}

export async function introduceNewCards(language: string, count_: number) {
  const session = await requireSession();
  const userId = session.user.id;

  const cards = await db
    .select()
    .from(srsCard)
    .where(
      and(
        eq(srsCard.userId, userId),
        eq(srsCard.language, language),
        eq(srsCard.status, "new")
      )
    )
    .orderBy(asc(srsCard.createdAt))
    .limit(count_);

  if (cards.length === 0) return [];

  const now = new Date();

  // Update all selected cards to learning status
  await db
    .update(srsCard)
    .set({
      status: "learning",
      nextReviewAt: now,
    })
    .where(
      and(
        eq(srsCard.userId, userId),
        eq(srsCard.language, language),
        eq(srsCard.status, "new"),
        inArray(
          srsCard.word,
          cards.map((c) => c.word)
        )
      )
    );

  return cards.map((c) => ({ ...c, status: "learning" as const, nextReviewAt: now }));
}

File: refactor_rust/backend/src/http/srs.rs
use axum::{
    Json,
    extract::{Path, Query, State},
};
use chrono::Utc;
use serde::Deserialize;
use serde_json::json;

use crate::{
    auth::AuthUser,
    db::models::SrsCard,
    error::{AppError, AppResult},
    srs::{CardStatus, ReviewInput, calculate_next_review},
    state::AppState,
};

#[derive(Deserialize)]
pub struct LanguageQuery {
    pub language: Option<String>,
    pub limit: Option<i64>,
}

pub async fn cards(
    State(state): State<AppState>,
    user: AuthUser,
    Query(q): Query<LanguageQuery>,
) -> AppResult<Json<Vec<SrsCard>>> {
    let limit = q.limit.unwrap_or(1000).clamp(1, 5000);
    let rows = if let Some(lang) = q.language {
        sqlx::query_as::<_, SrsCard>("SELECT * FROM srs_card WHERE user_id = $1 AND language = $2 ORDER BY created_at LIMIT $3")
            .bind(user.id()).bind(lang).bind(limit).fetch_all(&state.db).await?
    } else {
        sqlx::query_as::<_, SrsCard>(
            "SELECT * FROM srs_card WHERE user_id = $1 ORDER BY created_at LIMIT $2",
        )
        .bind(user.id())
        .bind(limit)
        .fetch_all(&state.db)
        .await?
    };
    Ok(Json(rows))
}

pub async fn stats(
    State(state): State<AppState>,
    user: AuthUser,
    Query(q): Query<LanguageQuery>,
) -> AppResult<Json<serde_json::Value>> {
    let lang = q.language;
    let (total, due, new_count, learning, review) = if let Some(lang) = lang {
        counts(&state, user.id(), Some(&lang)).await?
    } else {
        counts(&state, user.id(), None).await?
    };
    Ok(Json(
        json!({ "total": total, "due": due, "new": new_count, "learning": learning, "review": review, "learned": review }),
    ))
}

async fn counts(
    state: &AppState,
    user_id: &str,
    lang: Option<&str>,
) -> AppResult<(i64, i64, i64, i64, i64)> {
    let lang_filter = lang.unwrap_or("");
    let row = sqlx::query_as::<_, (i64, i64, i64, i64, i64)>(
        r#"SELECT
          count(*)::bigint,
          count(*) FILTER (WHERE status IN ('learning','review') AND next_review_at IS NOT NULL AND next_review_at <= now())::bigint,
          count(*) FILTER (WHERE status = 'new')::bigint,
          count(*) FILTER (WHERE status = 'learning')::bigint,
          count(*) FILTER (WHERE status = 'review')::bigint
        FROM srs_card WHERE user_id = $1 AND ($2 = '' OR language = $2)"#,
    )
    .bind(user_id).bind(lang_filter).fetch_one(&state.db).await?;
    Ok(row)
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AddWordBody {
    pub word: Option<String>,
    pub language: String,
    pub translation: Option<String>,
    pub words: Option<Vec<AddWordItem>>,
}
#[derive(Deserialize)]
pub struct AddWordItem {
    pub word: String,
    pub translation: String,
}

pub async fn add_word(
    State(state): State<AppState>,
    user: AuthUser,
    Path(word): Path<String>,
    Json(mut input): Json<AddWordBody>,
) -> AppResult<Json<serde_json::Value>> {
    input.word = Some(word);
    add_words(State(state), user, Json(input)).await
}

pub async fn add_words(
    State(state): State<AppState>,
    user: AuthUser,
    Json(input): Json<AddWordBody>,
) -> AppResult<Json<serde_json::Value>> {
    let mut items = input.words.unwrap_or_default();
    if let Some(word) = input.word {
        items.push(AddWordItem {
            word,
            translation: input.translation.unwrap_or_default(),
        });
    }
    if items.is_empty() {
        return Err(AppError::BadRequest("word is required".into()));
    }
    for item in &items {
        sqlx::query(
            "INSERT INTO srs_card (word, language, user_id, translation, status, next_review_at) VALUES ($1, $2, $3, $4, 'new', null) ON CONFLICT DO NOTHING",
        )
        .bind(item.word.to_lowercase()).bind(&input.language).bind(user.id()).bind(&item.translation)
        .execute(&state.db).await?;
    }
    Ok(Json(json!({ "success": true, "added": items.len() })))
}

pub async fn remove_word(
    State(state): State<AppState>,
    user: AuthUser,
    Path(word): Path<String>,
    Query(q): Query<LanguageQuery>,
) -> AppResult<Json<serde_json::Value>> {
    let lang = q
        .language
        .ok_or_else(|| AppError::BadRequest("language is required".into()))?;
    sqlx::query("DELETE FROM srs_card WHERE user_id = $1 AND language = $2 AND word = $3")
        .bind(user.id())
        .bind(lang)
        .bind(word.to_lowercase())
        .execute(&state.db)
        .await?;
    Ok(Json(json!({ "success": true })))
}

pub async fn remove_all_words(
    State(state): State<AppState>,
    user: AuthUser,
    Query(q): Query<LanguageQuery>,
) -> AppResult<Json<serde_json::Value>> {
    let lang = q
        .language
        .ok_or_else(|| AppError::BadRequest("language is required".into()))?;
    sqlx::query("DELETE FROM srs_card WHERE user_id = $1 AND language = $2")
        .bind(user.id())
        .bind(lang)
        .execute(&state.db)
        .await?;
    Ok(Json(json!({ "success": true })))
}

#[derive(Deserialize)]
pub struct ReviewBody {
    pub word: String,
    pub language: String,
    pub quality: i32,
}

pub async fn review(
    State(state): State<AppState>,
    user: AuthUser,
    Json(input): Json<ReviewBody>,
) -> AppResult<Json<serde_json::Value>> {
    let card = sqlx::query_as::<_, SrsCard>(
        "SELECT * FROM srs_card WHERE user_id = $1 AND language = $2 AND word = $3",
    )
    .bind(user.id())
    .bind(&input.language)
    .bind(input.word.to_lowercase())
    .fetch_optional(&state.db)
    .await?
    .ok_or(AppError::NotFound)?;
    let result = calculate_next_review(
        ReviewInput {
            ease_factor: card.ease_factor,
            interval: card.interval,
            repetitions: card.repetitions,
            status: CardStatus::from(card.status.as_str()),
        },
        input.quality,
    );
    sqlx::query("UPDATE srs_card SET ease_factor=$1, interval=$2, repetitions=$3, status=$4, next_review_at=$5, last_reviewed_at=$6 WHERE user_id=$7 AND language=$8 AND word=$9")
        .bind(result.ease_factor).bind(result.interval).bind(result.repetitions).bind(result.status).bind(result.next_review_at).bind(Utc::now().naive_utc())
        .bind(user.id()).bind(&input.language).bind(input.word.to_lowercase()).execute(&state.db).await?;
    Ok(Json(serde_json::to_value(result).unwrap()))
}

#[derive(Deserialize)]
pub struct IntroduceBody {
    pub language: String,
    pub count: i64,
}

pub async fn introduce_new(
    State(state): State<AppState>,
    user: AuthUser,
    Json(input): Json<IntroduceBody>,
) -> AppResult<Json<Vec<SrsCard>>> {
    let selected = sqlx::query_as::<_, SrsCard>("SELECT * FROM srs_card WHERE user_id=$1 AND language=$2 AND status='new' ORDER BY created_at LIMIT $3")
        .bind(user.id()).bind(&input.language).bind(input.count.clamp(1, 100)).fetch_all(&state.db).await?;
    for card in &selected {
        sqlx::query("UPDATE srs_card SET status='learning', next_review_at=now() WHERE user_id=$1 AND language=$2 AND word=$3")
            .bind(user.id()).bind(&input.language).bind(&card.word).execute(&state.db).await?;
    }
    let mut out = selected;
    for c in &mut out {
        c.status = "learning".into();
        c.next_review_at = Some(Utc::now().naive_utc());
    }
    Ok(Json(out))
}
